### PR TITLE
feat!: Migrate reference storage format to use new DXNs

### PR DESCRIPTION
**Pull request by dmaretskyi**
### TODO

- [x] Prevent indexer from reading old documents
- [x] SpaceProxy.createEpoch should wait for the new epoch and the new automerge root
- [x] Typenames are getting serialized to the wrong DXN kind: ` '@type': { '/': 'dxn:echo:@:dxos.org/type/Properties' },`
- [x] Make sure document version gets set for all new documents (roots & leafs)

### CODE REVIEW

File: packages/common/keys/src/dxn.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
+//
+// Copyright 2024 DXOS.org
+//
+
+import { invariant } from '@dxos/invariant';
+
+export class DXN {
```
Comment (wittjosiah) [inline packages/common/keys/src/dxn.ts:7]: would be good to include jsdoc on the class and include the basic spec

File: packages/common/keys/src/dxn.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    if (!(typeof kind === 'string' && kind.length > 0)) {
+      throw new Error('Invalid DXN');
+    }
+    if (!(parts.length > 0)) {
+      throw new Error('Invalid DXN');
+    }
+    return new DXN(kind, parts);
+  }
+
+  #kind: string;
```
Comment (wittjosiah) [inline packages/common/keys/src/dxn.ts:30]: not seen this hash syntax before, what does this do?

File: packages/core/echo/echo-db/src/serializer.ts [hunk 3]
```diff
@@ -155,21 +166,28 @@ export class Serializer {
       ...data,
       '@version': Serializer.version,
       '@meta': meta,
-      '@timestamp': new Date().toUTCString(),
+      '@timestamp': new Date().toISOString(),
```
Comment (wittjosiah) [inline packages/core/echo/echo-db/src/serializer.ts:169]: is this just the timestamp of when it was serialized?

File: packages/core/echo/echo-protocol/src/space-doc-version.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
+//
+// Copyright 2024 DXOS.org
+//
+
+/**
+ * Denotes the data version of the space document.
```
Comment (wittjosiah) [inline packages/core/echo/echo-protocol/src/space-doc-version.ts:6]: document here referring to "automerge document"?

File: packages/core/echo/indexing/src/index-schema.ts [hunk 5]
```diff
@@ -4,7 +4,7 @@
 
 import { Event } from '@dxos/async';
 import { Resource } from '@dxos/context';
-import { type ObjectStructure } from '@dxos/echo-protocol';
+import { decodeReference, type ObjectStructure } from '@dxos/echo-protocol';
 import { EXPANDO_TYPENAME } from '@dxos/echo-schema';
```
Comment (wittjosiah) [inline packages/core/echo/indexing/src/index-schema.ts:8]: does this typename need to be updated as well?

File: packages/common/keys/src/dxn.ts (hunk 2 above)
Comment (dmaretskyi) [reply packages/common/keys/src/dxn.ts:30]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_properties

File: packages/core/echo/indexing/src/index-schema.ts (hunk 5 above)
Comment (dmaretskyi) [reply packages/core/echo/indexing/src/index-schema.ts:8]: cc @richburdon 

Currently it is:

```ts
export const EXPANDO_TYPENAME = 'Expando';
```

Comment (wittjosiah) [reply packages/core/echo/indexing/src/index-schema.ts:8]: presumably it should be `dxos.org/type/Expando` and we might as well do that at the same time as the space properties

File: packages/apps/types/src/serializer/serializers/default.ts [hunk 6]
```diff
@@ -87,7 +87,7 @@ const deserializeEchoObject = (parsed: any, newId?: boolean): Expando => {
     core.id = id;
   }
   getMeta(deserializedObject).keys.push(...(meta?.keys ?? []));
-  const typeRef = getTypeRef(type);
+  const typeRef = internalDecodeReference(type);
```
Comment (wittjosiah) [inline packages/apps/types/src/serializer/serializers/default.ts:90]: maybe todo to remove usage? I think this was the only external usage of the api?

File: packages/gravity/proto-guard/src/check-snapshot.test.ts [hunk 7]
```diff
@@ -30,35 +31,59 @@ describe('Load client from storage snapshot', () => {
     return testStoragePath;
   };
 
-  test('check if space loads for Automerge on nodeFS snapshot', async () => {
-    const snapshot = SnapshotsRegistry.getSnapshot('automerge-nodeFS');
-    invariant(snapshot, 'Snapshot not found');
-    log.info('Testing snapshot', { snapshot });
+  test
+    .skip('check if space loads for Automerge on nodeFS snapshot', async () => {
```
Comment (wittjosiah) [inline packages/gravity/proto-guard/src/check-snapshot.test.ts:35]: why is this test being skipped?

File: packages/sdk/client/src/echo/space-proxy.ts [hunk 8]
```diff
@@ -435,6 +451,12 @@ export class SpaceProxy implements Space {
     });
   }
 
+  private async _migrate() {
```
Comment (wittjosiah) [inline packages/sdk/client/src/echo/space-proxy.ts:454]: why this vs space.internal.migrate?

File: packages/gravity/proto-guard/src/check-snapshot.test.ts (hunk 7 above)
Comment (dmaretskyi) [reply packages/gravity/proto-guard/src/check-snapshot.test.ts:35]: its getting removed in https://github.com/dxos/dxos/pull/7031

File: packages/sdk/client/src/echo/space-proxy.ts (hunk 8 above)
Comment (dmaretskyi) [reply packages/sdk/client/src/echo/space-proxy.ts:454]: This is exposed as `space.internal.migrate`

**Pull request merged**